packages/langium/src/references/linker.ts

```diff
@@ -5,26 +5,40 @@
  ******************************************************************************/
 
 import { LangiumServices } from '../services';
-import { AstNode, AstReflection } from '../syntax-tree';
-import { ScopeProvider } from './scope';
+import { AstNode } from '../syntax-tree';
+import { AstNodeDescription, ScopeProvider } from './scope';
 
 export interface Linker {
     link(node: AstNode, referenceName: string, referenceId: string): AstNode | undefined;
+    // TODO should be a collection of AstNodeDescriptions?
+    linkingCandiates(node: AstNode, referenceName: string, referenceId: string): AstNodeDescription | undefined;
 }
 
 export class DefaultLinker implements Linker {
     protected readonly scopeProvider: ScopeProvider;
-    protected readonly reflection: AstReflection;
+    protected readonly services: LangiumServices;
 
     constructor(services: LangiumServices) {
+        this.services = services;
         this.scopeProvider = services.references.ScopeProvider;
-        this.reflection = services.AstReflection;
     }
 
     link(node: AstNode, referenceName: string, referenceId: string): AstNode | undefined {
+        const description = this.linkingCandiates(node, referenceName, referenceId);
+        if (description)
+            return this.loadAstNode(description);
+        return undefined;
```

Comment:
We discussed that the linker should be responsible of creating Reference objects, so the logic in the `buildReference` method in LangiumParser should be moved here.